classes/observation_manager.php

```diff
@@ -0,0 +1,402 @@
+            );
+        }
+
+        $transaction->allow_commit();
+    }
+
+    /**
+     * Reorders an observation point in relation to the other observation points in this observation instance
+     * @param int $observationid ID of the observation instance
+     * @param int $obpointid ID of the observation point to reorder
+     * @param int $direction direction and magnitude to reorder the point in
+     * @param string $tablename database table name
+     */
+    public static function reorder_observation_point(int $observationid, int $obpointid, int $direction,
+            string $tablename='observation_points') {
+        global $DB;
+
+        if (!is_int($direction) || $direction == 0) {
+            throw new \coding_exception("direction must be an integer that is not zero");
+        }
+
+        // First get the ordering of the current point.
+        $targetpoint = self::get_existing_point_data($observationid, $obpointid, $tablename);
+
+        // Get all the points to get the bounds.
+        $allpoints = self::get_observation_points($observationid);
+        $alllistorders = array_column($allpoints, 'list_order');
+
+        if (count($alllistorders) === 0) {
+            throw new \coding_exception('No list orderings found for this observation instance, but expected at least one.');
+        }
+
+        $newordering = $targetpoint->list_order + $direction;
+
+        // Is currently the minimum - do nothing.
+        if ($newordering < min($alllistorders)) {
+            return;
+        }
+
+        // Is currently the maximum - do nothing.
+        if ($newordering > max($alllistorders)) {
+            return;
+        }
+
+        // Ordering is valid, so get the points that are affected by this reordering.
+        $affectedpoints = array_filter($allpoints, function($elem) use($newordering, $direction, $targetpoint) {
+            // Don't include the item being reordered.
+            if ($elem->id === $targetpoint->id) {
+                return false;
+            }
+
+            // If shifting down, filter those 'above'.
+            if ($direction < 0) {
+                return $elem->list_order >= $newordering && $elem->list_order < $targetpoint->list_order;
+            }
+            // If shifting up, filter those 'below'.
+            if ($direction > 0) {
+                return $elem->list_order <= $newordering && $elem->list_order > $targetpoint->list_order;
+            }
+
+            return false;
+        });
+
+        $transaction = $DB->start_delegated_transaction();
+
+        // Give the target point the new ordering.
+        $DB->update_record($tablename,
+        [
+            'id' => $targetpoint->id,
+            'list_order' => $newordering
+        ]);
+
+        // Reduce the direction to a unit vector (e.g. 5 -> 1 and -5 -> -1).
+        $reductionamount = intdiv($direction, abs($direction));
+
+        array_walk($affectedpoints, function($elem) use($DB, $tablename, $reductionamount) {
```

Comment:
This could likely be simpler just using a for loop here, rather than an array walk, which is a little more similar in concept to an array_map